lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,752 @@
+  template <typename... Args>
+  BinaryExprBase(Args &&... args)
+      : BinaryExpr(DerivedKind, std::forward<Args>(args)...) {}
+  static bool classof(const Expr *e) { return e->kind == DerivedKind; }
+};
+
+/// An addition.
+struct AddExpr : public BinaryExprBase<AddExpr, Expr::KindAdd> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "(" << lhs() << " + " << rhs() << ")";
+  }
+};
+
+/// The maximum of two expressions.
+struct MaxExpr : public BinaryExprBase<MaxExpr, Expr::KindMax> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "max(" << lhs() << ", " << rhs() << ")";
+  }
+};
+
+/// The minimum of two expressions.
+struct MinExpr : public BinaryExprBase<MinExpr, Expr::KindMin> {
+  using BinaryExprBase::BinaryExprBase;
+  void print(llvm::raw_ostream &os) const {
+    os << "min(" << lhs() << ", " << rhs() << ")";
+  }
+};
+
+void Expr::print(llvm::raw_ostream &os) const {
+  TypeSwitch<const Expr *>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { e->print(os); });
+}
+
+Expr::iterator Expr::begin() const {
+  return TypeSwitch<const Expr *, Expr::iterator>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { return e->begin(); });
+}
+
+Expr::iterator Expr::end() const {
+  return TypeSwitch<const Expr *, Expr::iterator>(this)
+      .Case<RootExpr, VarExpr, KnownExpr, AddExpr, MaxExpr, MinExpr>(
+          [&](auto *e) { return e->end(); });
+}
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// GraphTraits on constraint expressions
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct GraphTraits<Expr *> {
+  using ChildIteratorType = Expr::iterator;
+  using NodeRef = Expr *;
+
+  static NodeRef getEntryNode(NodeRef node) { return node; }
+  static inline ChildIteratorType child_begin(NodeRef node) { return node->begin(); }
+  static inline ChildIteratorType child_end(NodeRef node) { return node->end(); }
+};
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Fast bump allocator with optional interning
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// An allocation slot in the `InternedAllocator`.
+template <typename T>
+struct InternedSlot {
+  T *ptr;
+  InternedSlot(T *ptr) : ptr(ptr) {}
+};
+
+/// A simple bump allocator that ensures only ever one copy per object exists.
+/// The allocated objects must not have a destructor.
+template <typename T, typename std::enable_if_t<
+                          std::is_trivially_destructible<T>::value, int> = 0>
+class InternedAllocator {
+  using Slot = InternedSlot<T>;
+  llvm::DenseSet<Slot> interned;
+  llvm::BumpPtrAllocator allocator;
```

Comment:
Putting these graph nodes into a bump pointer allocator is a great idea, thank you for doing this.  Should these three different allocators all share the same one though?  The lifetime is all tied together IIUC.

Reply:
That's an excellent idea, done!